tlatools/org.lamport.tlatools/test/pcal/parser/PcalParserTranslator.java

```diff
@@ -0,0 +1,184 @@
+	 * 
+	 * @param input The unparsed PlusCal text.
+	 * @return A parsed PlusCal Abstract Syntax Tree (AST).
+	 * @throws ParseAlgorithmException If the PlusCal syntax is invalid.
+	 */
+	public static AST parse(String input) throws ParseAlgorithmException {
+		Vector<String> lines = new Vector<String>(Arrays.asList(input.split("\r?\n")));
+		PcalStart start = new PcalStart(lines);
+
+		// Some static PlusCal parser initialization
+		TLAtoPCalMapping mapping = new TLAtoPCalMapping() ;
+		mapping.algLine = start.line;
+		mapping.algColumn = start.column;
+		PcalParams.tlaPcalMapping = mapping;
+
+		// Parse algorithm
+		int endLine = lines.size();
+		int endColumn = (int)lines.lastElement().codePoints().count();
+		PcalCharReader reader = new PcalCharReader(lines, start.line, start.column, endLine, endColumn);
+		return ParseAlgorithm.getAlgorithm(reader, start.isFair);
+	}
+	
+	/**
+	 * Translates the given TLA+ token. We just hardcode a small set of
+	 * tokens used as expression placeholders in the test corpus.
+	 * 
+	 * @param token The TLA+ token to translate.
+	 * @return The corresponding node in normalized form.
+	 */
+	public static AstNode translate(TLAToken token) {
+		switch (token.string) {
+			case "TRUE": {
+				return Kind.BOOLEAN.asNode();
+			} default: {
+				return Kind.IDENTIFIER_REF.asNode();
+			}
+		}
+	}
+	
+	/**
+	 * Translates the given TLA+ expression. The PlusCal parser does very
+	 * limited actual parsing of TLA+ expressions, simply identifying their
+	 * extents then providing a list of lines of tokens. So we keep TLA+
+	 * expressions quite simple here. It could be interesting in the future
+	 * to call into SANY to parse these, so we can analyze the logic that
+	 * determines expression extent.
+	 * 
+	 * @param expr The TLA+ expression, as a sequence of lines of tokens.
+	 * @return The corresponding node in normalized form.
+	 */
+	public static AstNode translate(TLAExpr expr) {
+		Assert.assertEquals(1, expr.tokens.size()); // Lines of tokens
+		Assert.assertEquals(1, expr.tokens.get(0).size()); // Tokens on a given line
+		return translate(expr.tokens.get(0).get(0));
+	}
+	
+	/**
+	 * Translates the PlusCal AST generated by the parser into a normalized
+	 * form for comparison to the unified parser test corpus.
+	 * 
+	 * TODO: This translation is largely incomplete in its current form. The
+	 * bones are now in place for a comprehensive PlusCal syntax corpus test
+	 * similar to what exists for SANY, but now somebody needs to do the real
+	 * work of translating every syntax node to its universal corpus form.
+	 * 
+	 * @param astNode A node in the PlusCal AST.
+	 * @return The corresponding node in normalized universal corpus form.
+	 * @throws ParseAlgorithmException If a construct could not be translated.
+	 */
+	public static AstNode translate(AST astNode) throws ParseAlgorithmException {
+		if (astNode.getClass().equals(AST.UniprocessObj.getClass())) {
```

Comment:
I respect that this matches how the translator does case analysis on AST nodes, but I don't think there's any reason to use the `AST.*Obj` fields in new code. Just use e.g. `AST.Uniprocess.class` instead.